source/extensions/filters/udp/udp_proxy/session_filters/filter.h

```diff
@@ -60,10 +60,21 @@ enum class ReadFilterStatus {
   StopIteration,
 };
 
+class FilterBase {
+public:
+  virtual ~FilterBase() = default;
+
+  /**
+   * This routine is called before the access log handlers' final log() is called. Filters can use
+   * this callback to enrich the data passed in to the log handlers.
+   */
+  virtual void onSessionComplete() {}
```

Comment:
Can you add a bool in this class and use the bool to make sure onSessionComplete() is only called once instead of relying on is_read_write_filter_ in its wrapper? Something like:

```
void onSessionComplete() {
   if (already_called) return;
   onSessionCompleteInternal();
}

protected:
virtual void onSessionCompleteInternal() {
ASSERT(!already_called);
}
```